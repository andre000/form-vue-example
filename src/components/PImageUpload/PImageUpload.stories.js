import '@/assets/scss/storybook.scss'
import { withKnobs } from '@storybook/addon-knobs'
import PImageUpload from './PImageUpload.vue'
import image from './__mocks__/image'

export default {
  component: PImageUpload,
  title: 'PImageUpload',
  decorator: [withKnobs]
}

export const withInitialState = () => ({
  components: { PImageUpload },
  template: '<p-image-upload />'
})

export const withPreview = () => ({
  components: { PImageUpload },
  props: {
    image: {
      default: image
    }
  },
  template: '<p-image-upload :image="{ preview: image }"/>'
})
