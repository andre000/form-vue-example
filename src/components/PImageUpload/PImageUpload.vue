<template>
  <div
    class="image-upload"
    @click="handleClick"
    @dragover.stop.prevent="handleDrag"
    @drop.stop.prevent="handleDrop"
  >
    <input
      ref="image-order"
      type="file"
      accept=".png, .jpg, .jpeg"
      name="image-order"
      class="image-upload__input"
      @change="handleUpload"
    />

    <icon-upload v-if="!image" class="image-upload__icon hide-sm hide-xs" />
    <span v-if="!image" class="image-upload__message">
      Jogue aqui o arquivo de imagem do seu pastel ou clique para localizar a
      pasta.
    </span>

    <img v-if="image && image.preview" :src="image.preview" alt="" />
  </div>
</template>

<script>
import iconUpload from '@/assets/img/upload.svg'

export default {
  name: 'PImageUpload',
  components: {
    iconUpload
  },
  props: {
    image: null
  },

  watch: {
    image() {
      if (this.image === null) {
        this.$refs['image-order'].value = ''
      }
    }
  },

  methods: {
    handleClick() {
      this.$refs['image-order'].click()
    },

    handleDrag() {},

    handleDrop(event) {
      this.$refs['image-order'].files = event.dataTransfer.files
      this.handleUpload({ target: { files: event.dataTransfer.files } })
    },

    handleUpload(e) {
      if (!e.target.files.length) return
      this.generatePreview()
    },

    generatePreview() {
      const reader = new FileReader()
      reader.onload = e => {
        this.$emit('upload', {
          preview: e.target.result,
          file: this.$refs['image-order'].files[0]
        })
      }
      reader.readAsDataURL(this.$refs['image-order'].files[0])
    }
  }
}
</script>

<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';

.image-upload {
  min-width: 100px;
  border: 1px solid $primary-text;
  border-radius: 10px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  text-align: center;
  cursor: pointer;
  padding: 20px;
  position: relative;
  background: #fff;

  .image-upload__input {
    display: none;
  }

  .image-upload__icon {
    height: 45px;
  }

  .image-upload__message {
    margin-top: 5px;
    font: 16px/21px Roboto;
    color: $primary-text;
  }

  img {
    object-fit: contain;
    width: calc(100% + 40px);
    height: 100%;
    border-radius: 10px;
  }

  &:hover {
    .upload-up-arrow {
      animation: updown 0.7s ease-in-out infinite alternate;
    }
  }
}

@keyframes updown {
  25% {
    transform: translateY(0px);
  }
  100% {
    transform: translateY(2px);
  }
}
</style>
