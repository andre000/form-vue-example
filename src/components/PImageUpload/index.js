export { default as PImageUpload } from './PImageUpload.vue'
