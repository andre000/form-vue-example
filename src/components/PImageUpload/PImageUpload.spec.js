import { shallowMount, createLocalVue } from '@vue/test-utils'

import PImageUpload from './PImageUpload'

const localVue = createLocalVue()

describe('PImageUpload component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(PImageUpload, {
      localVue
    })
  })

  it('should match the snapshot', () =>
    expect(wrapper.html()).toMatchSnapshot())

  it('should have an input[file]', () => {
    expect(wrapper.find('input[type="file"]').exists()).toBe(true)
  })

  it.only('should generate preview when a file is uploaded', async () => {
    const mockedFile = [
      {
        name: 'image.png',
        size: 50000,
        type: 'image/png'
      }
    ]

    const readAsDataURLSpy = jest
      .spyOn(FileReader.prototype, 'readAsDataURL')
      .mockImplementation(function() {
        this.onload({ target: { result: 'foo' } })
      })

    const input = wrapper.find('input[type="file"]')
    wrapper.vm.handleUpload({ target: { files: mockedFile } })
    await input.trigger('change')

    expect(wrapper.emitted().upload).toBeTruthy()
    await wrapper.setProps({ image: wrapper.emitted().upload[0][0] })

    const preview = wrapper.find('img')
    expect(readAsDataURLSpy).toHaveBeenCalled()
    expect(preview.exists()).toBe(true)
  })
})
