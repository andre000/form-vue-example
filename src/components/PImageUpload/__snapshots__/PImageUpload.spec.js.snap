// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PImageUpload component should match the snapshot 1`] = `
<div class="image-upload"><input type="file" accept=".png, .jpg, .jpeg" name="image-order" class="image-upload__input">
  <!---->
</div>
`;
